### PR TITLE
ENH: Update version of DCMTK to 3.6.1_201610.12

### PR DESCRIPTION
**Pull request by che85**

### CODE REVIEW

**Comment by fedorov** [comment]
@che85 this depends on patched ITK here, correct?

https://github.com/fbudin69500/ITK/tree/update_DCMTK

**Comment by che85** [comment]
Yes, correct.

**Comment by che85** [comment]
Wait, no I didn't update the revision tag from the External.cmake here. So actually this PR doesn't depend on the patched ITK...

**Comment by fedorov** [comment]
Right, but CTK reused ITK from Slicer, and that ITK is the patched one. CTK does depend on ITK, so if you were to compile it standalone without updating ITK hash in CTK externals, I would expect CTK build to fail.

**Comment by che85** [comment]
I will check this out.

**Comment by che85** [comment]
@fedorov You were right. It doesn't build with the currently defined ITK revision in https://github.com/commontk/CTK/blob/master/CMakeExternals/ITK.cmake#L28

Should I update this or just wait for further action of Francois?

**Comment by fedorov** [comment]
Additional fixes will need to be done to ITK to fix Slicer issues you identified, so things will change.

I would not spend any more time on this until we hear back from @thewtex and @fbudin69500.

**Comment by fbudin69500** [comment]
This looks good to me but a decision has to be made on the DCMTK commit to choose for the update:
18e2abbd930f00fb41c1cea42cec04ccf0cf8637 (08/23/2016)
efe1a177eac040b6d41aae0db7714cd970e12bc5 (10/12/2016)

**Comment by fbudin69500** [comment]
@jcfr  @thewtex  Since Slicer and CTK do not require C++11 features, this patch is good to be merged.

File: CMakeExternals/DCMTK.cmake [hunk 1]
```diff
@@ -24,7 +24,7 @@ if(DEFINED DCMTK_DIR AND NOT EXISTS ${DCMTK_DIR})
 endif()
 
 if(NOT DEFINED DCMTK_DIR AND NOT ${CMAKE_PROJECT_NAME}_USE_SYSTEM_${proj})
-  set(revision_tag DCMTK-3.6.1_20150924)
+  set(revision_tag efe1a177eac040b6d41aae0db7714cd970e12bc5)
```
Comment (jcfr) [inline CMakeExternals/DCMTK.cmake]: Could you add a comment specifying what is this version ? Is it the latest nightly that was available ?

Comment (jcfr) [inline CMakeExternals/DCMTK.cmake]: Or even better,  should the tag `3.6.1_201610.12` be used ?

Comment (fbudin69500) [inline CMakeExternals/DCMTK.cmake]: There is no such tag in DCMTK upstream (http://git.dcmtk.org/dcmtk)

Comment (che85) [inline CMakeExternals/DCMTK.cmake]: Exactly. That's why I added a comment in the a new commit.

Comment (fbudin69500) [inline CMakeExternals/DCMTK.cmake]: @che85 thanks!

**Comment by jcfr** [comment]
Thanks for the patches.

Next will be the update of Slicer ?

**Comment by fbudin69500** [comment]
yes :)

**Comment by fedorov** [comment]
Note that we are waiting for some additional merges to DCMTK, and possibly release of the next snapshot early November (@michaelonken is coordinating this). If/when this happens, we will need to go through the update process again.

**Comment by thewtex** [comment]
Nice work all!

**Comment by nolden** [comment]
@fbudin69500 why does it matter whether Slicer or CTK uses C++ 11 features?

**Comment by jcfr** [comment]
> why does it matter whether Slicer or CTK uses C++ 11 features? 

@nolden  At that point in time, it doesn't really for CTK and Slicer. That said, some folks that depends on ITK which depends on DCMTK build their project (e.g BRAINSTools maintained by @hjmjohnson) are using C++11, this is why this came up in the discussion. @fbudin69500 was doing some testing build BRAINSTools and found the issue.

The good news is that @michaelonken mentioned that they work working on fixing DCMTK.

Et voila, hope this clarify things.

**Comment by nolden** [comment]
Ok, thanks @jcfr . MITK already uses C++11 features and passes 
CMAKE_CXX_STANDARD:STRING=11 to all external projects in the SuperBuild. 
No problems so far but good to know before the next CTK/DCMTK update

**Comment by fbudin69500** [comment]
@nolden DCMTK requires the CMake variable boolean "DCMTK_USE_CXX11_STL" to specify that it needs to be built with C++11 support. My guess is that when it is built with MITK, it is actually not build with C++11 features. Let me know if I am mistaken.

**Comment by hjmjohnson** [comment]
New version of DCMTK still does not compile with C++11

 

In file included from /scratch/johnsonhj/src/BT-11/DCMTK/ofstd/tests/tvariant.cc:5:

/scratch/johnsonhj/src/BT-11/DCMTK/ofstd/include/dcmtk/ofstd/ofvriant.h:61:9: warning: type definition in a constexpr function

      is a C++14 extension [-Wc++14-extensions]

        struct test { alignas(OFvariant_traits<Index+1,Alternatives...>::alignment()) first_alternative a; };

        ^

/scratch/johnsonhj/src/BT-11/DCMTK/ofstd/include/dcmtk/ofstd/ofvriant.h:61:23: error: requested alignment is less than minimum

      alignment of 4 for type 'first_alternative' (aka 'float')

        struct test { alignas(OFvariant_traits<Index+1,Alternatives...>::alignment()) first_alternative a; };

                      ^

/scratch/johnsonhj/src/BT-11/DCMTK/ofstd/include/dcmtk/ofstd/ofvriant.h:61:16: note: in instantiation of member class 'test'

      requested here

        struct test { alignas(OFvariant_traits<Index+1,Alternatives...>::alignment()) first_alternative a; };

               ^

/scratch/johnsonhj/src/BT-11/DCMTK/ofstd/include/dcmtk/ofstd/ofvriant.h:61:74: note: in instantiation of member function

      'OFvariant_traits<1, float, bool>::alignment' requested here

        struct test { alignas(OFvariant_traits<Index+1,Alternatives...>::alignment()) first_alternative a; };

                                                                         ^

/scratch/johnsonhj/src/BT-11/DCMTK/ofstd/include/dcmtk/ofstd/ofvriant.h:61:16: note: in instantiation of member class 'test'

      requested here

        struct test { alignas(OFvariant_traits<Index+1,Alternatives...>::alignment()) first_alternative a; };

               ^

/scratch/johnsonhj/src/BT-11/DCMTK/ofstd/include/dcmtk/ofstd/ofvriant.h:316:21: note: in instantiation of member function

      'OFvariant_traits<0, int, float, bool>::alignment' requested here

    alignas(traits::alignment()) std::uint8_t m_Content[traits::size()];

                    ^

/scratch/johnsonhj/src/BT-11/DCMTK/ofstd/tests/tvariant.cc:60:31: note: in instantiation of template class

      'OFvariant<int, float, bool>' requested here

    OFvariant<int,float,bool> v0( true );

                              ^

1 warning and 1 error generated.

**Pull request merged**